src/foam/nanos/iban/ValidationIBAN.js

```diff
@@ -150,98 +150,275 @@ foam.CLASS({
+        return iban;
+      },
+      javaCode: `
+        int l = iban.length();
+        for ( int i = l - l % 4 ; i > 0 ; i -= 4 ) iban = iban.substring(0, i) + ' ' + iban.substring(i);
+        return iban;
+      `
     },
-
-    function getFormatForCountry(cc) {
-      var cInfo = this.COUNTRIES[cc];
-      return cInfo && cInfo[0];
+    {
+      name: 'getFormatForCountry',
+      type: 'String',
+      args: [
+        {
+          name: 'cc',
+          type: 'String'
+        }
+      ],
+      code:  function(cc) {
+        var cInfo = this.COUNTRIES[cc];
+        return cInfo && cInfo[0];
+      },
+      javaCode: `
+        Object[] cInfo = (Object[]) COUNTRIES.get(cc);
+        if ( cInfo != null ) return (String) cInfo[0];
+        return "null";
+      `
     },
+    {
+      name: 'validate',
+      type: 'Void',
+      args: [
+        {
+          name: 'iban',
+          type: 'String'
+        }
+      ],
+      code: function(iban) {
+        iban = this.trim(iban);
+
+        if ( iban == '' ) return `${this.IBAN_REQUIRED}`;
 
-    function validate(iban) {
-      iban = this.trim(iban);
+        let cc = iban.substring(0, 2);
+        let format = this.getFormatForCountry(cc);
 
-      if ( iban == '' ) return `${this.IBAN_REQUIRED}`;
+        if ( ! format ) return `${this.UNKNOWN_COUNTRY_CODE} ${cc}`;
 
-      var cc = iban.substring(0, 2);
-      var format = this.getFormatForCountry(cc);
+        if ( iban.length != format.length + 2 )
+          return `${this.LENGTH_MISMATCHED_1} ${(format.length + 2)} ${this.LENGTH_MISMATCHED_2} ${iban.length}`;
 
-      if ( ! format ) return `${this.UNKNOWN_COUNTRY_CODE} ${cc}`;
+        for ( let i = 0 ; i < format.length ; i++ ) {
+          if ( ! this.validateChar(format.charAt(i), iban.charAt(i+2)) ) return `${this.INVALID_CHARACTER}  '${(i+2)}'`;
+        }
 
-      if ( iban.length != format.length + 2 )
-        return `${this.LENGTH_MISMATCHED_1} ${(format.length + 2)} ${this.LENGTH_MISMATCHED_2} ${iban.length}`;
+        let num = this.toNumber(iban);
+        let checksum = this.mod(num, 97);
 
-      for ( var i = 0 ; i < format.length ; i++ ) {
-        if ( ! this.validateChar(format.charAt(i), iban.charAt(i+2)) ) return `${this.INVALID_CHARACTER} ${(i+2)}`;
-      }
+        if ( checksum != 1 ) return `${this.INVALID_CHECKSUM} ${checksum}`;
 
-      var num = this.toNumber(iban);
-      var checksum = this.mod(num, 97);
+        return 'passed';
+      },
+      javaCode: `
+        iban = trim(iban);
+        if ( "".equals(iban) ) throw new IllegalStateException(IBAN_REQUIRED);
 
-      if ( checksum != 1 ) return `${this.INVALID_CHECKSUM} ${checksum}`;
+        String cc = iban.substring(0, 2);
+        String format = getFormatForCountry(cc);
 
-      return 'passed';
-    },
+        if ( format == null ) throw new IllegalStateException(UNKNOWN_COUNTRY_CODE + cc);
```

Comment:
The call getFormatForCountry presently returns "null" rather than null.

Reply:
Throw foam.core.ValidationException rather than IllegalStateException, here and elsewhere.

Reply:
"null" - too much following js code, will fix.

nanopay/bank/BRBankAccount.js expects IllegalStateException from all validations, should I change this as well?

Reply:
really, that was unexpected.  yes, that would be nice.